bip380/miniscript/property.py

```diff
@@ -59,21 +61,26 @@ def check_valid(self):
                 num_types += 1
 
         # Check for conflicts in type & properties.
-        if not (
-            (not self.z or not self.o)
-            and (not self.n or not self.z)
-            and (not self.V or not self.d)
-            and (not self.K or self.u)
-            and (not self.V or not self.u)
-            and (not self.e or not self.f)
-            and (not self.e or self.d)
-            and (not self.V or not self.e)
-            and (not self.d or not self.f)
-            and (not self.V or self.f)
-            and (not self.K or self.s)
-            and (not self.z or self.m)
-        ):
-            raise MiniscriptPropertyError("Conflicting types and properties")
+
+        checks = [
+            lambda: (not self.z or not self.o),
+            lambda: (not self.n or not self.z),
+            lambda: (not self.V or not self.d),
+            lambda: (not self.K or self.u),
+            lambda: (not self.V or not self.u),
+            lambda: (not self.e or not self.f),
+            lambda: (not self.e or self.d),
+            lambda: (not self.V or not self.e),
+            lambda: (not self.d or not self.f),
+            lambda: (not self.V or self.f),
+            lambda: (not self.K or self.s),
+            lambda: (not self.z or self.m),
+        ]
+        conflicts = [fn for fn in checks if not fn()]
+        if conflicts:
+            formatted_conflicts = " ".join([inspect.getsource(fn).strip().replace("self.", "").replace("lambda: ", "")
+                                            for fn in conflicts])[:-1]
+            raise MiniscriptPropertyError(f"Conflicting types and properties: {formatted_conflicts}")
```

Comment:
Nice hack with `inspect` :-). But i think it would be cleaner if we leveraged the string representation here? For instance (untested):
```diff
diff --git a/bip380/miniscript/property.py b/bip380/miniscript/property.py
index 2f30b05..fa0c251 100644
--- a/bip380/miniscript/property.py
+++ b/bip380/miniscript/property.py
@@ -60,27 +60,27 @@ class Property:
                     )
                 num_types += 1
 
-        # Check for conflicts in type & properties.
-
-        checks = [
-            lambda: (not self.z or not self.o),
-            lambda: (not self.n or not self.z),
-            lambda: (not self.V or not self.d),
-            lambda: (not self.K or self.u),
-            lambda: (not self.V or not self.u),
-            lambda: (not self.e or not self.f),
-            lambda: (not self.e or self.d),
-            lambda: (not self.V or not self.e),
-            lambda: (not self.d or not self.f),
-            lambda: (not self.V or self.f),
-            lambda: (not self.K or self.s),
-            lambda: (not self.z or self.m),
+        # Check for conflicts in type & properties. For each type checked, give
+        # the properties it must have and the one it must not have.
+        conflicts = [
+            ("K", "", "us"),
+            ("V", "de", "f"),
+            ("z", "o", "m"),
+            ("n", "z", ""),
+            ("e", "d", "f"),
+            ("d", "", "f"),
         ]
-        conflicts = [fn for fn in checks if not fn()]
-        if conflicts:
-            formatted_conflicts = " ".join([inspect.getsource(fn).strip().replace("self.", "").replace("lambda: ", "")
-                                            for fn in conflicts])[:-1]
-            raise MiniscriptPropertyError(f"Conflicting types and properties: {formatted_conflicts}")
+        for (typ, must_be, must_not_be) in conflicts:
+            if not getattr(self, typ):
+                continue
+            if not all(getattr(self, prop) for prop in must_be):
+                raise MiniscriptPropertyError(
+                    f"Conflicting types and properties: '{typ}' needs always be '{must_be}'"
+                )
+            if any(getattr(self, prop) for prop in must_not_be):
+                raise MiniscriptPropertyError(
+                    f"Conflicting types and properties: '{typ}' can't be any of '{must_not_be}'"
+                )
 
     def type(self):
         return "".join(filter(lambda x: x in self.types, str(self)))
```